v2/internal/reconcilers/azuresql/connector.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package sql
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	ctrlconversion "sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	asosql "github.com/Azure/azure-service-operator/v2/api/sql/v1"
+	sql "github.com/Azure/azure-service-operator/v2/api/sql/v1api20211101/storage"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/conditions"
+)
+
+type Connector interface {
+	CreateOrUpdate(ctx context.Context) error
+	Delete(ctx context.Context) error
+	Exists(ctx context.Context) (bool, error)
+}
+
+type ownerDetails struct {
+	fqdn     string
+	database string
+}
+
+func getOwnerDetails(ctx context.Context, resourceResolver *resolver.Resolver, user *asosql.User) (ownerDetails, error) {
+	// Get the owner - at this point it must exist
+	owner, err := resourceResolver.ResolveOwner(ctx, user)
+	if err != nil {
+		return ownerDetails{}, err
+	}
+
+	hierarchy, err := resourceResolver.ResolveResourceHierarchy(ctx, owner.Owner)
+	if err != nil {
+		return ownerDetails{}, err
+	}
+
+	if len(hierarchy) != 3 {
+		return ownerDetails{}, errors.Errorf("failed to look up ownerDetails: expected resource hierarchy len=3 but was %d", len(hierarchy))
+	}
+
+	genericServer := hierarchy[1]
+	genericDatabase := hierarchy[2]
+
+	server, ok := genericServer.(*sql.Server)
+	if !ok {
+		return ownerDetails{}, errors.Errorf("owner's owner was not type Server, instead: %T", genericServer)
+	}
+
+	database, ok := genericDatabase.(*sql.ServersDatabase)
+	if !ok {
+		return ownerDetails{}, errors.Errorf("owner was not type ServersDatabase, instead: %T", genericDatabase)
+	}
+
+	// Magical assertion to ensure that this is still the storage type
```

Comment:
Suggest being explicit about what's needed if this doesn't compile.

```suggestion
	// Assertion to ensure that this is still the storage type
	// If this doesn't compile, update the version being imported to the new Hub version
```

Reply:
This comment was actually copied from elsewhere. I improved here and also the other 2 instances of it.